test: fix flaky tests `Dapp viewed Event...` `Permissions Page...` and `Connections Page...` due to overlaying popup

seaona opened the pull request:
## **Description**
This is a PR for fixing the flaky tests:
- Dapp viewed Event @no-mmi is sent when reconnect to a dapp that has been connected before
- Permissions Page should redirect users to connections page when users click on connected permission
- Connections page should disconnect when click on Disconnect button in connections page

The problem seems to be that we are trying to click an element, but there is the previous notifcation element on top and has not disappeared yet at the moment we want to perform the click operation.

We can see this in the artifacts of the failure, where the blue notification popup is still there, despite having already clicked 'Got it'.
This is also confirmed by the error message in the logs:

```
[driver] Called 'clickElement' with arguments [{"text":"Got it","tag":"button"}]
[driver] Called 'clickElement' with arguments [{"text":"127.0.0.1:8080","tag":"p"}]
Failure on testcase: 'Dapp viewed Event @no-mmi is sent when reconnect to a dapp that has been connected before', for more information see the artifacts tab in CI
ElementClickInterceptedError: Element <p class="mm-box mm-text mm-text--body-md mm-text--elllipsis mm-text--text-align-left mm-box--color-text-default">
 is not clickable at point (444,92) because another element <h4 class="mm-box mm-text mm-text--heading-sm mm-text--text-align-center mm-box--width-full mm-box--color-info-inverse"> obscures it
```

The solution here is to wait for the element to disappear. Since this seems a case that can happen more often, a new driver function has been introduced, which basically clicks an element and waits until it disappears. This can be suitable for when accepting notifications/popup that whenever we accept them, those should disappear before we proceed with the next action.

```
  async clickElementAndWaitToDisappear(rawLocator) {
    const element = await this.findClickableElement(rawLocator);
    await element.click();
    await element.waitForElementState('hidden');
  }
 ```

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24575 and https://github.com/MetaMask/metamask-extension/issues/24597

## **Manual testing steps**

1. locally: run the test locally several times ` yarn test:e2e:single test/e2e/tests/metrics/dapp-viewed.spec.js --browser=chrome --leave-running=true`
2. ci: Check ci and see is not failing

## **Screenshots/Recordings**
Artifacts and logs from ci that helped in the investigation:

![image](https://github.com/MetaMask/metamask-extension/assets/54408225/a63bea82-1a8b-49bc-8037-c389a36579fe)

![Screenshot from 2024-05-17 11-27-45](https://github.com/MetaMask/metamask-extension/assets/54408225/5fdef2d3-8049-445b-a795-a1d7a6f16959)


## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

seaona committed:
fix flaky test for dapp viewed event when reconnect to a dapp

legobeat commented on test/e2e/webdriver/driver.js:
```diff
@@ -373,6 +373,12 @@ class Driver {
+  async clickElementAndWaitToDisappear(rawLocator) {
+    const element = await this.findClickableElement(rawLocator);
+    await element.click();
+    await element.waitForElementState('hidden');
```
Would it make sense to set an explicit timeout here of the maximum time we expect the notification element to be visible in this step?

seaona replied:
absolutely @legobeat :bowing_woman: :heart:  I've pushed a new commit to address that

seaona committed:
add default timeout and js description

legobeat reviewed (dismissed):
LGTM!

legobeat committed:
Merge branch 'develop' into flaky-fix-dapp-viewed

seaona committed:
test: fix flaky test `Permissions Page ...` and ` Connections Page...` (#24607)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR fixes the following flaky tests:

- Permissions Page should redirect users to connections page when users
click on connected permission
- Connections page should disconnect when click on Disconnect button in
connections page

After this PR is merged, for fixing this tests we use the same approach,
utilizing the new method for waiting for the notification to disappear
after accepting it

```
async clickElementAndWaitToDisappear(rawLocator) {
    const element = await this.findClickableElement(rawLocator);
    await element.click();
    await element.waitForElementState('hidden');
  }
```

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24607?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24597

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**
Artifacts from ci


![image](https://github.com/MetaMask/metamask-extension/assets/54408225/e0bdfeab-8fec-4201-b447-e65dd552ebc7)


<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

seaona commented:
@legobeat there were more instances of this type of flakyiness, so I've decided to add the fixes in this same PR. Apologies for resetting the review, I just saw it now in this ticket 
:pray: :bowing_woman: 
https://github.com/MetaMask/metamask-extension/issues/24597

legobeat committed:
Merge branch 'develop' into flaky-fix-dapp-viewed

chloeYue reviewed (approved):
LGTM! Thanks !
